using Microsoft.AspNetCore.Http;
using PersonApi.Entities.Exceptions;
using PersonApi.Models.Exceptions;
using System;
using System.Net;
using System.Threading.Tasks;

namespace PersonApi.Middleware
{
    public class ErrorHandlingMiddleware
    {
        private readonly RequestDelegate _next;

        public ErrorHandlingMiddleware(RequestDelegate next)
        {
            _next = next;
        }

        public async Task Invoke(HttpContext context)
        {
            try
            {
                await _next(context);
            }
            catch(InternalException e)
            {
                if(!context.Response.HasStarted)
                {
                    context.Response.ContentType = "application/json";
                    context.Response.StatusCode = (int)HttpStatusCode.BadRequest;
                }

                await context.Response.WriteAsync(e.Message);
            }
            catch(NotFoundException e)
            {
                if (!context.Response.HasStarted)
                {
                    context.Response.ContentType = "application/json";
                    context.Response.StatusCode = (int)HttpStatusCode.NotFound;
                }

                await context.Response.WriteAsync(e.Message);
            }
            catch(Exception e)
            {
                if (!context.Response.HasStarted)
                {
                    context.Response.ContentType = "application/json";
                    context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
                }

                await context.Response.WriteAsync(e.Message);
            }
        }
    }
}
